packages/react-composites/src/composites/common/Composite.tsx

```diff
@@ -23,9 +23,15 @@ export interface BaseCompositeProps {
   /**
    * Locale for the composite.
    *
-   * @defaultValue English (US)
+   * @defaultValue English (US) locale
    */
-  locale?: CompositeLocale;
+  locale?: CompositeLocale | ((locale?: string) => Promise<CompositeLocale>);
```

Comment:
Similar to my comment about `LocalizationProvider` vs `AsyncLocalizationProvider`, it is better to keep the two ways to set locales separate, since they are very different things, rather than a single field that we type-switch on later on.

Reply:
I wonder how we should expose the Composite API to cover both usages `LocalizationProvider` and `AsyncLocalizationProvider`. Two props doesn't (locale and localeLoader) doesn't feel right because both props can be provided.

Reply:
I think tagged unions are a good fit for this:

```ts

// Replace with actual type
type CompositeLocale = string;

interface AsyncLocaleLoader {
  kind: 'asyncLoader';
  loader: () => Promise<CompositeLocale>;
  /**
   * If set, the locale to use before locale can be loaded via `loader`.
   * 
   * If unset, en-US locale strings included with this library will be used.
   */
  defaultLocale?: CompositeLocale;
}

interface StaticLocale {
  kind: 'staticValue';
  value: CompositeLocale;
}

type LocaleArgument = AsyncLocaleLoader | StaticLocale;

const newComposite = (arg: LocaleArgument) => {
  switch(arg.kind) {
    case 'asyncLoader':
      console.log(`This is async loader with arguments ${arg.defaultLocale} and ${arg.loader}`);
      break;
    case 'staticValue':
      console.log(`This is static value with arguments ${arg.value}`);
  }
}
```
[playground](https://www.typescriptlang.org/play?#code/FAehAICUFMAcBsCGBja4DuBLALgC3CtgK6LzjYCes0wl14AwgPYC2sTAzjtADJPKk0AXnAdsAJ0wA7AOYBuYMGnZo4gGYo0AQQ4UpyPgPi8miACarwAb2DhwAa2lmAXOADkiXfr7nVbhXbwphbirgAUAJTgQgB84AAK4qyYHNAAPMxsnNyGgjEB4CAAVEW2dkXgAJJqotDYADTkuGhBRmjYTOBEqeAARtBqTOIt-ILgAlJ9I75m4ABumIjgAAZBvuLLAHRl4BU7FdVdUqkN4NBSALQAqgDK4K1jYpKyHODSyPBEFrNYeE0p90wvXEiHEFAwmHgZH6XVSZm2dl2IDKFg0RHg2FyxgA-K5MuwuCosdAFABfRTKVQaVDgG7YRDYTAGUbGaxlRxSFzuMQMpkANVIRGg-jKc0F0DxrAJORZJOA5NoVDQxK04hkRBY52w0XAOj0zLaPhC4AAPrT6YyDYIFMBkExjtqpNB0PjsiodWFQTJXCq1RqtVFYmy7BxfshcJ61ZsOWYojZEeNPGgPF5mes3M4doi7ccmMZNkEZGFlgAVXAAgGefX3YKWX74L3+qTYV4AEisXs2qMQ6MxstJBE54HbnbWIVJywiBQTwOgiHs08TPTcPMtAs+wszCbsOY4eegBaYRdL5deANXTPm4ohf0bmubbY7UbFG4nU7K5NJQA)

Reply:
I tried this but it changed the usage in the composite usage to something like this:
```
<ChatComposite adapter={chatAdapter} locale={{ kind: 'asyncLoader', loader: () => COMPOSITE_LOCALE_FR_FR, defaultLocale: COMPOSITE_LOCALE_EN_US}} />}
<CallComposite adapter={callAdapter} locale={{ kind: 'staticValue', value: COMPOSITE_LOCALE_JA_JP}} />
```
I found this to be less delightful.
How about two props:
```
  /**
   * Locale for the composite.
   *
   * @defaultValue English (US) locale
   */
  locale?: CompositeLocale;
  /**
   * Asynchronous callback to change locale
   *
   * @defaultValue undefined
   */
  localeLoader?: () => Promise<CompositeLocale>;
```
Such that when `localeLoader` is defined then we use the `AsyncLocalizationProvider` with `locale` as the default locale. If `localeLoader` is not defined then `LocalizationProvider` is used with `locale`.